### PR TITLE
[openstack_instack] Collect undercloud debug information

### PR DESCRIPTION
**Pull request by stuggi**
If auth information got sourced before sosreport run
(OS_USERNAME, OS_PASSWORD, OS_TENANT_NAME) collect
information from nova, neutron, keystone, glance, ironic
and heat.
Only undercloud specific heat information is collected in
instack plugin.

Signed-off-by: Martin Schuppert <mschuppe@redhat.com>

### CODE REVIEW

File: sos/plugins/openstack_glance.py [hunk 1]
```diff
@@ -44,6 +46,20 @@ def setup(self):
...
+                self.soslog.warning("%s not found in environment variables"
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
+                run_cmds = True
+
+        if run_cmds:
+            self.add_cmd_output(
+                "openstack image list --long",
+                suggest_filename="openstack_image_list--long")
```
Comment (bmr-cymru) [inline sos/plugins/openstack_glance.py]: Don't set `suggest_filename` unless you are setting it to something other than the default mangled command name.

File: sos/plugins/openstack_heat.py [hunk 2]
```diff
@@ -32,6 +34,20 @@ def setup(self):
...
+                self.soslog.warning("%s not found in environment variables"
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
+                run_cmds = True
+
+        if run_cmds:
+            self.add_cmd_output(
+                "openstack stack list",
+                suggest_filename="openstack_stack_list")
```
Comment (bmr-cymru) [inline sos/plugins/openstack_heat.py]: Same thing for `suggest_filename` here.

File: sos/plugins/openstack_instack.py [hunk 3]
```diff
@@ -28,6 +30,41 @@ def setup(self):
...
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
+                run_cmds = True
+
+        if run_cmds:
+            # get status of overcloud stack and resources
+            self.add_cmd_output(
+                "openstack stack show overcloud",
+                suggest_filename="openstack_stack_show_overcloud")
```
Comment (bmr-cymru) [inline sos/plugins/openstack_instack.py]: Same thing for `suggest_filename` here.

File: sos/plugins/openstack_instack.py [hunk 4]
```diff
@@ -28,6 +30,41 @@ def setup(self):
...
+            self.add_cmd_output(
+                "openstack stack resource list -n 10 overcloud",
+                suggest_filename="openstack_stack_resource_list-n10_overcloud")
```
Comment (bmr-cymru) [inline sos/plugins/openstack_instack.py]: Same thing for `suggest_filename` here.

File: sos/plugins/openstack_instack.py [hunk 5]
```diff
@@ -28,6 +30,41 @@ def setup(self):
...
+                if 'FAILED' in deployment:
+                    check = [
+                        "OS::Heat::StructuredDeployment",
+                        "OS::Heat::SoftwareDeployment"]
+                    if any(x in deployment for x in check):
+                        deployment = deployment.split()[1]
+                        cmd = "openstack software deployment show " \
+                                "--long %s" % (deployment)
+                        self.add_cmd_output(
+                            cmd,
```
Comment (bmr-cymru) [inline sos/plugins/openstack_instack.py]: Why does this need to be on a newline? Only lists inside `[..]` use the bracket-then-newline style in sos unless there's no space on the line.

If that makes the line with the `kwarg` harder to format I'd suggest using a variable for the file name - building them in-line is always a bit tricky to format readably since the argument name is already so long.

File: sos/plugins/openstack_ironic.py [hunk 6]
```diff
@@ -37,6 +39,23 @@ def setup(self):
...
+                self.soslog.warning("%s not found in environment variables"
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
+                run_cmds = True
+
+        if run_cmds:
+            self.add_cmd_output(
+                "openstack baremetal node list --long",
+                suggest_filename="openstack_baremetal_node_list--long")
```
Comment (bmr-cymru) [inline sos/plugins/openstack_ironic.py]: Same thing for `suggest_filename` here.

File: sos/plugins/openstack_ironic.py [hunk 7]
```diff
@@ -37,6 +39,23 @@ def setup(self):
...
+            self.add_cmd_output(
+                "openstack baremetal port list",
+                suggest_filename="openstack_baremetal_port_list")
+
```
Comment (bmr-cymru) [inline sos/plugins/openstack_ironic.py]: Same thing for `suggest_filename` here.

File: sos/plugins/openstack_keystone.py [hunk 8]
```diff
@@ -41,6 +43,23 @@ def setup(self):
...
+                self.soslog.warning("%s not found in environment variables"
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
+                run_cmds = True
+
+        if run_cmds:
+            self.add_cmd_output(
+                "openstack endpoint list",
+                suggest_filename="openstack_endpoint_list")
```
Comment (bmr-cymru) [inline sos/plugins/openstack_keystone.py]: Same thing for `suggest_filename` here.

File: sos/plugins/openstack_keystone.py [hunk 9]
```diff
@@ -41,6 +43,23 @@ def setup(self):
...
+            self.add_cmd_output(
+                "openstack catalog list",
+                suggest_filename="openstack_catalog_list")
```
Comment (bmr-cymru) [inline sos/plugins/openstack_keystone.py]: Same thing for `suggest_filename` here.

File: sos/plugins/openstack_neutron.py [hunk 10]
```diff
@@ -36,6 +38,26 @@ def setup(self):
...
+                self.soslog.warning("%s not found in environment variables"
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
+                run_cmds = True
+
+        if run_cmds:
+            self.add_cmd_output(
+                "openstack subnet list",
+                suggest_filename="openstack_subnet_list")
```
Comment (bmr-cymru) [inline sos/plugins/openstack_neutron.py]: Same thing for `suggest_filename` here.

Comment (bmr-cymru) [inline sos/plugins/openstack_neutron.py]: (and the next two in this file)

**Comment by bmr-cymru** [comment]
pep8 no likey:

```
$ pep8 sos
sos/plugins/openstack_instack.py:62:33: E127 continuation line over-indented for visual indent
```

**Comment by stuggi** [comment]
Thanks, yes I should have not set explicitly the file names . I think I have addressed them all. Also added mistral and zaqar to the instack plugin, but missed to mention this in the commit message. Let me know if I should change the commit message.

File: sos/plugins/openstack_glance.py [hunk 11]
```diff
@@ -44,6 +46,18 @@ def setup(self):
...
         self.add_copy_spec("/etc/glance/")
 
+        run_cmds = False
+        for os_var in ['OS_USERNAME', 'OS_PASSWORD', 'OS_TENANT_NAME']:
+            if os_var not in os.environ:
+                self.soslog.warning("%s not found in environment variables"
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
+                run_cmds = True
```
Comment (pmoravec) [inline sos/plugins/openstack_glance.py]: Do you want to set run_cmds to True if and only if *all* the three variables are set? If so, then this wont work.. Check os.environ containing just one of the three vars - you get two warnings but run_cmds=True.

(or is your intent to set run_cmds if at least one such variable is set? that's what the code does..)

File: sos/plugins/openstack_heat.py [hunk 12]
```diff
@@ -32,6 +34,18 @@ def setup(self):
             suggest_filename="heat_db_version"
         )
 
+        run_cmds = False
+        for os_var in ['OS_USERNAME', 'OS_PASSWORD', 'OS_TENANT_NAME']:
+            if os_var not in os.environ:
+                self.soslog.warning("%s not found in environment variables"
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
```
Comment (pmoravec) [inline sos/plugins/openstack_heat.py]: Ditto as above. If the intention is to set run_cmds to True only if all three vars are set, then try e.g.:

    run_cmds = True
    for os_var ..:
        if os_var not in os.environ:
            run_cmds = False
            self.soslog.warning(..)

File: sos/plugins/openstack_instack.py [hunk 13]
```diff
@@ -28,6 +30,50 @@ def setup(self):
...
+            self.add_copy_spec_limit("/var/log/zaqar/*.log",
+                                     sizelimit=self.limit)
+
+        run_cmds = False
+        for os_var in ['OS_USERNAME', 'OS_PASSWORD', 'OS_TENANT_NAME']:
+            if os_var not in os.environ:
+                self.soslog.warning("%s not found in environment variables"
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
```
Comment (pmoravec) [inline sos/plugins/openstack_instack.py]: ditto as above

File: sos/plugins/openstack_ironic.py [hunk 14]
```diff
@@ -37,6 +39,19 @@ def setup(self):
 
         self.add_cmd_output('ls -laRt /var/lib/ironic/')
 
+        run_cmds = False
+        for os_var in ['OS_USERNAME', 'OS_PASSWORD', 'OS_TENANT_NAME']:
+            if os_var not in os.environ:
+                self.soslog.warning("%s not found in environment variables"
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
```
Comment (pmoravec) [inline sos/plugins/openstack_ironic.py]: ditto

File: sos/plugins/openstack_keystone.py [hunk 15]
```diff
@@ -41,6 +43,19 @@ def setup(self):
             self.add_copy_spec_limit("/var/log/keystone/*.log",
                                      sizelimit=self.limit)
 
+        run_cmds = False
+        for os_var in ['OS_USERNAME', 'OS_PASSWORD', 'OS_TENANT_NAME']:
+            if os_var not in os.environ:
+                self.soslog.warning("%s not found in environment variables"
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
```
Comment (pmoravec) [inline sos/plugins/openstack_keystone.py]: ditto

File: sos/plugins/openstack_neutron.py [hunk 16]
```diff
@@ -36,6 +38,20 @@ def setup(self):
         self.add_copy_spec("/etc/neutron/")
         self.add_copy_spec("/var/lib/neutron/")
 
+        run_cmds = False
+        for os_var in ['OS_USERNAME', 'OS_PASSWORD', 'OS_TENANT_NAME']:
+            if os_var not in os.environ:
+                self.soslog.warning("%s not found in environment variables"
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
```
Comment (pmoravec) [inline sos/plugins/openstack_neutron.py]: ditto

File: sos/plugins/openstack_nova.py [hunk 17]
```diff
@@ -28,17 +29,18 @@ class OpenStackNova(Plugin):
...
-            self.add_cmd_output(
-                "nova service-list",
-                suggest_filename="nova_service_list")
+        run_cmds = False
+        for os_var in ['OS_USERNAME', 'OS_PASSWORD', 'OS_TENANT_NAME']:
+            if os_var not in os.environ:
+                self.soslog.warning("%s not found in environment variables"
+                                    " which is required to run commands"
+                                    % (os_var))
+            else:
```
Comment (pmoravec) [inline sos/plugins/openstack_nova.py]: ditto

File: sos/plugins/openstack_glance.py (hunk 11 above)
Comment (bmr-cymru) [inline sos/plugins/openstack_glance.py]: You can also avoid the explicit loop entirely by using a list comprehension and `all()`:
```python
vars = [p in os.environ for p in ['OS_USERNAME', 'OS_PASSWORD', 'OS_TENANT_NAME']]
if not all(vars):
    # freak out
```

Comment (stuggi) [inline sos/plugins/openstack_glance.py]: Oh yes, I blindly c/p it from the existed nova plugin. I'll fix this and submit an updated patch. Also I will check the other openstack plugins we already have. I know e.g. the same is in gnocchi.

**Comment by stuggi** [comment]
changes implemented. also submitted a fix for gnocchi in https://github.com/sosreport/sos/pull/956

File: sos/plugins/openstack_glance.py [hunk 18]
```diff
@@ -2,6 +2,7 @@
 # Copyright (C) 2012 Rackspace US, Inc.,
```
Comment (bmr-cymru) [inline sos/plugins/openstack_glance.py]: The commit subject is misleading - if you are going to change all the OpenStack plugins in one patch, use `[openstack_*]` as the commit tag. Otherwise you should never e.g. have hunks that apply to `openstack_glance.py` in a patch with a subject of `[openstack_instack]` - it's just confusing for anyone trying to understand the logs down the line.

Your other option here would be to make one commit per file - this is especially preferred if you are making other sets of unrelated changes (e.g. the very large addition to `openstack_insight.py` vs. the relatively minore env. var. change in all the others).

I'm not going to make you respin and split up the patch series at this stage, but it is the preferred style in `sos` (and many other projects), but the commit subject definitely needs fixing before this can go in.

File: sos/plugins/openstack_nova.py [hunk 19]
```diff
@@ -28,44 +29,37 @@ class OpenStackNova(Plugin):
...
+            self.add_cmd_output("openstack hypervisor stats show")
+            # get details for each nova instance
+            cmd = "openstack server list -f value"
+            nova_instances = self.call_ext_prog(cmd)['output']
+            for instance in nova_instances.splitlines():
+                instance = instance.split()[0]
+                cmd = "openstack server show %s" % (instance)
+                self.add_cmd_output(
+                    cmd,
+                    suggest_filename="instance-" + instance + ".log")
```
Comment (bmr-cymru) [inline sos/plugins/openstack_nova.py]: Ditto these changes - would have been much better to do the code shuffling in a separate patch.

File: sos/plugins/openstack_glance.py (hunk 18 above)
Comment (stuggi) [inline sos/plugins/openstack_glance.py]: will split in furute

**Comment by stuggi** [comment]
changed commit subject to openstack_* . Will split such change in future to different patches.

**Comment by bmr-cymru** [comment]
Thanks! I'll get these into this week's review batch.

**Comment by bmr-cymru** [comment]
Cheers - these look fine now. Will hit master in a moment.
